README.md
# Pushex.js

[![npm](https://img.shields.io/npm/v/pushex.svg)](https://npmjs.com/package/pushex)
[![Hex.pm](https://img.shields.io/hexpm/l/push_ex.svg)](https://github.com/pushex-project/pushex.js/blob/master/LICENSE)
[![Build Status](https://travis-ci.org/pushex-project/pushex.js.svg?branch=master)](https://travis-ci.org/pushex-project/pushex.js)

[PushEx](https://github.com/pushex-project/pushex) is an Elixir based push-only websocket implementation. Pushex.js is the
JS client to the server.

## Links

- [Hexdocs.pm](https://hexdocs.pm/push_ex)
- [Pushex Github](https://github.com/pushex-project/pushex)
- [Pushex.js npm](https://www.npmjs.com/package/pushex)

## Installation

You can install Pushex.js by using [npm](https://www.npmjs.com/package/pushex):

```
npm install --save pushex
```

## Documentation ToC

- [Pushex#constructor/2](#pushexconstructor)
- [Pushex#connect/0](#pushexconnect)
- [Pushex#disconnect/0](#pushexdisconnect)
- [Pushex#resetParams/0](#pushexresetparams)
- [Pushex#subscribe/1](#pushexsubscribe)
- [Pushex#unsubscribe/1](#pushexunsubscribe)
- [Pushex#getExistingSubscription/1](#pushexgetexistingsubscription)
- [Subscription#bind/2](#subscriptionbind)
- [Subscription#unbind/1](#subscriptionunbind)
- [Subscription#unbindAll/0](#subscriptionunbindall)
- [Subscription#hasBindings/0](#subscriptionhasbindings)

## Usage

Pushex is exposed via an exported class `Pushex`:

```js
import { Pushex } from 'pushex'
```

This class can be used to create an instance of Pushex. Each instance of Pushex maintains its own state and does not share any state with other Pushex instances. Thus, it's possible to run multiple Pushex clients on the same webpage.

```js
const pushex = new Pushex('wss://example.myserver.com/push_socket', {})
pushex.subscribe('user-1').bind('*', (event, data) => {
  console.log('user-1 channel received event/data', event, data)
})
```

Pushex constructor can accept several optional parameters that are important for extending the library for any use case:

```js
const pushex = new Pushex('wss://example.myserver.com/push_socket', {
  getParams: () => Promise.resolve({ token: 'myToken' }),
  onConnect: (pushEx) => {},
  onConnectionError: (pushEx) => {},
  socketReconnectAlgorithm: (tries) => {
    return [3000, 6000, 10000, 20000][tries - 1] || 30000
  }
})
```

## Authentication

All authentication / authorization occurs on your hosted server running PushEx during the lifecycle points `socket_connect` and `channel_join`.

## Methods

All public methods are detailed below.

### Pushex#constructor

```js
constructor(url, { getParams, onConnect, onConnectionError, socketReconnectAlgorithm })
```

* getParams - Function that returns a promise resolving parameters passed to the server on connection. This is only ever invoked in `connect` or explicit using `resetParams`.
* onConnect - This is invoked when the socket is opened
* onConnectionError - This is invoked when the socket encounters an error
* socketReconnectAlgorithm - This is invoked when the socket is determining how it should back off. The return value is in `ms`. The default implementation is `[3000, 6000, 10000, 20000][tries - 1] || 30000` and does not need specified

### Pushex#connect

```js
connect()
```

Resolves connection params and connects to the push socket. If the socket is already connected, the socket will **not** be disconnected.

### Pushex#disconnect

```js
disconnect()
```

Disconnects the push socket.

### Pushex#resetParams

```js
resetParams()
```

Fetches params using the provided `getParams` function and sets the params on the socket. This will not change the connected socket, but will affect future connection attempts, including reconnection attempts.

### Pushex#subscribe

```js
subscribe(channelName)
```

This is the main interface to accessing push data. A Subscription will be created and memoized for future usage. The `channelName` is used to connect as a Phoenix channel. In PushEx, any channel name you want to use is valid, and the authorization layer can be used to deny invalid channels.

### Pushex#unsubscribe

```js
unsubscribe(channelName)
```

If a Subscription exists for this name it will be left gracefully and closed. The subscription is removed from the list of PushEx subscriptions so that calling `subscribe/1` again will create a new Subscription. This resolves a promise because you are not guaranteed the Subscription is closed until a message is delivered to the server.

### Pushex#getExistingSubscription

```js
getExistingSubscription(channelName)
```

Returns the subscription if it already exists. Does not create a new subscription.

### Subscription#bind

```js
bind(eventName, callbackFn)
```

A subscription channel can have events bound to it, with the callbackFn being invoked when a message is received from the server.

In order to capture all events on a given cahnnel, use the special channelName `'*'`. The callbackFn will be invoked on every channel event, regardless of the name.

The `callbackFn` is invoked as `callbackFn(data, event)`.

`bind` returns a function which can be used to unsubscribe the event:

```js
const unsub = pushex.subscribe('myCh').bind('*', myCallback)
// myCallback is invoked when myCh receives any event
unsub()
// myCallback will not be invoked through this callback
```

### Subscription#unbind

```js
unbindAll(eventName)
```

All event bindings for the provided name will be removed. The Subscription still remains open and alive, so you could rebind events and it would work.

### Subscription#unbindAll

```js
unbindAll()
```

All event bindings for the entire subscription will be removed. The Subscription still remains open and alive, so you could rebind events and it would work.

### Subscription#hasBindings

```js
hasBindings()
```

Returns true/false indicating whether there are active bindings for this subscription.


CHANGELOG.md
# Change Log

This project adheres to [Semantic Versioning](http://semver.org/).

Every release, along with the migration instructions, is documented on the Github [Releases page](https://github.com/pushex-project/pushex.js/releases).

## 0.2.0

This PR updates Phoenix JS to the latest 1.5.13 version. No changes otherwise.

## 0.1.0

This PR includes a switch over to TypeScript. Of course, there is now an index.d.ts file for you to pull from!
There are no breaking changes nor new code features, other than native TS support.

Thanks to @raygesualdo for this overhaul!


src/index.spec.js
import { Pushex } from "./index"

it("exports Pushex", () => {
  expect(Pushex).toBeDefined()
})


src/pushex.spec.js
import { Socket } from "phoenix"

import { Pushex } from "./pushex"

it("defines a Phoenix Socket to the current URL", () => {
  const pushex = new Pushex("wss://test.com", {})
  expect(pushex.socket).toEqual(expect.any(Socket))
  expect(pushex.socket.endPoint).toEqual("wss://test.com/websocket")
})

it("uses a default reconnectAfterMs strategy", () => {
  const pushex = new Pushex("wss://test.com", {})
  expect(pushex.socket.reconnectAfterMs(1)).toEqual(3000)
  expect(pushex.socket.reconnectAfterMs(2)).toEqual(6000)
  expect(pushex.socket.reconnectAfterMs(3)).toEqual(10000)
  expect(pushex.socket.reconnectAfterMs(4)).toEqual(20000)
  expect(pushex.socket.reconnectAfterMs(5)).toEqual(30000)
  expect(pushex.socket.reconnectAfterMs(6)).toEqual(30000)
  expect(pushex.socket.reconnectAfterMs(7)).toEqual(30000)
})

it("invokes onConnect when the socket becomes connected", (done) => {
  const getParams = () => Promise.resolve({ a: 1 })
  const onConnect = () => {
    done()
  }

  const pushex = new Pushex("wss://test.com", { onConnect, getParams })
  // Simulate a socket conn open, but setup the socket first
  pushex.connect().then(() => {
    pushex.socket.conn.onopen()
  })
})

it("invokes onConnectionError when the socket encounters an error", (done) => {
  const getParams = () => Promise.resolve({ a: 1 })
  const onConnectionError = () => {
    done()
  }

  const pushex = new Pushex("wss://test.com", { onConnectionError, getParams })
  // Simulate a socket conn error, but setup the socket first
  pushex.connect().then(() => {
    pushex.socket.conn.onerror()
  })
})

describe("connect", () => {
  it("fetches params before connecting the socket, if it is not already connected", () => {
    let counter = 0
    const getParams = () => {
      return Promise.resolve({ counter: counter++ })
    }

    const pushex = new Pushex("wss://test.com", { getParams })
    return pushex.connect().then(() => {
      expect(pushex.socket.conn.url).toEqual("wss://test.com/websocket?counter=0&vsn=2.0.0")

      return pushex.disconnect().then(() => {
        return pushex.connect().then(() => {
          expect(pushex.socket.conn.url).toEqual("wss://test.com/websocket?counter=1&vsn=2.0.0")
        })
      })
    })
  })

  it("does not change the socket params if the socket was already connected", () => {
    let counter = 0
    const getParams = () => {
      return Promise.resolve({ counter: counter++ })
    }

    const pushex = new Pushex("wss://test.com", { getParams })
    return pushex.connect().then(() => {
      expect(pushex.socket.conn.url).toEqual("wss://test.com/websocket?counter=0&vsn=2.0.0")

      return pushex.connect().then(() => {
        expect(pushex.socket.conn.url).toEqual("wss://test.com/websocket?counter=0&vsn=2.0.0")
      })
    })
  })
})

describe("disconnect", () => {
  it("disconnects the socket", () => {
    const pushex = new Pushex("wss://test.com", {})
    return pushex.connect().then(() => {
      expect(pushex.socket.conn).toBeInstanceOf(window.WebSocket)

      return pushex.disconnect().then(() => {
        expect(pushex.socket.conn).toEqual(null)
      })
    })
  })
})

describe("resetParams", () => {
  it("fetches and sets params on the socket, without changing the connected socket", () => {
    let counter = 0
    const getParams = () => {
      return Promise.resolve({ counter: counter++ })
    }

    const pushex = new Pushex("wss://test.com", { getParams })
    return pushex.connect().then(() => {
      expect(pushex.socket.conn.url).toEqual("wss://test.com/websocket?counter=0&vsn=2.0.0")
      expect(pushex.socket.params()).toEqual({ counter: 0 })

      return pushex.resetParams().then(() => {
        expect(pushex.socket.params()).toEqual({ counter: 1 })
        expect(pushex.socket.conn.url).toEqual("wss://test.com/websocket?counter=0&vsn=2.0.0")
      })
    })
  })
})

describe("subscribe", () => {
  it("defines, sets up, and returns a Subscription instance", () => {
    const pushex = new Pushex("wss://test.com", {})
    const sub = pushex.subscribe("chName")

    expect(sub.channelName).toEqual("chName")
    expect(sub.pushEx).toEqual(pushex)
    expect(sub.channel).not.toBeUndefined()
  })

  it("can be called multiple times to return the same Subscription instance", () => {
    const pushex = new Pushex("wss://test.com", {})
    const sub = pushex.subscribe("chName")
    const sameSub = pushex.subscribe("chName")
    const sub2 = pushex.subscribe("other")

    expect(sub).toEqual(sameSub)
    expect(sub).not.toEqual(sub2)
  })
})

describe("unsubscribe", () => {
  it("resolves when there isn't an existing subscription", (done) => {
    const pushex = new Pushex("wss://test.com", {})
    pushex.unsubscribe("chName").then(() => {
      done()
    })
  })

  it("closes an existing subscription", (done) => {
    const pushex = new Pushex("wss://test.com", {})
    const sub = pushex.subscribe("chName")

    expect(pushex.subscriptions.chName).not.toEqual(undefined)
    expect(sub.channel).not.toEqual(undefined)

    pushex.unsubscribe("chName").then(() => {
      expect(pushex.subscriptions).toEqual({})
      expect(sub.channel).toEqual(undefined)
      done()
    })
  })
})


src/subscription.spec.js
import { Subscription } from "./subscription"

let mockChannel, mockSocket, mockPushEx

beforeEach(() => {
  mockChannel = {
    join: jest.fn(),
    on: jest.fn(),
    leave: jest.fn(),
  }

  mockSocket = {
    channel: jest.fn(() => mockChannel),
  }

  mockPushEx = {
    getSocket: jest.fn(() => mockSocket),
  }
})

describe("setup", () => {
  it("creates/joins a channel through the socket", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    expect(subscription.setup()).toEqual(subscription)

    expect(mockSocket.channel.mock.calls).toEqual([["chName"]])
    expect(mockChannel.join.mock.calls).toEqual([[]])
  })

  it("sets up the msg callback of the channel", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    expect(mockChannel.on.mock.calls).toEqual([["msg", expect.any(Function)]])
  })
})

describe("msg handler", () => {
  it("runs all bindings for the event", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const args = []
    let called = ""
    subscription.bind("event", (data, event) => {
      args.push([data, event])((called += "a"))
    })

    subscription.bind("event", () => (called += "b"))
    subscription.bind("eventx", () => fail("should not match"))

    mockChannel.on.mock.calls[0][1]({ data: "data", event: "event" })
    expect(args).toEqual([["data", "event"]])
    expect(called).toEqual("ab")

    mockChannel.on.mock.calls[0][1]({ data: "data", event: "event" })
    expect(args).toEqual([
      ["data", "event"],
      ["data", "event"],
    ])
    expect(called).toEqual("abab")

    mockChannel.on.mock.calls[0][1]({ data: "data", event: "nope" })
    expect(called).toEqual("abab")
  })

  it("runs * bindings as all messages", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const args = []
    let called = ""
    subscription.bind("event", () => (called += "a"))
    subscription.bind("*", (data, event) => {
      args.push([data, event])((called += "w"))
    })
    subscription.bind("eventx", () => fail("should not match"))

    mockChannel.on.mock.calls[0][1]({ data: "data", event: "event" })
    expect(args).toEqual([["data", "event"]])
    expect(called).toEqual("aw")

    mockChannel.on.mock.calls[0][1]({ data: "data", event: "event" })
    expect(called).toEqual("awaw")

    mockChannel.on.mock.calls[0][1]({ data: "data", event: "only the wildcard" })
    expect(called).toEqual("awaww")
  })

  it("swallows errors into a timeout to allow future subscriptions to be invoked while preserving the error stack", () => {
    jest.useFakeTimers()
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    let called = ""
    subscription.bind("event", () => (called += "a"))
    subscription.bind("event", () => {
      throw new Error("x")
    })
    subscription.bind("event", () => (called += "b"))

    mockChannel.on.mock.calls[0][1]({ data: "fake", event: "event" })

    expect(called).toEqual("ab")
    expect(() => {
      jest.runAllTimers()
    }).toThrowError(new Error("x"))
  })

  it("can remove bindings", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    let called = ""
    const unsubA = subscription.bind("event", () => (called += "a"))
    const unsubB = subscription.bind("event", () => (called += "b"))

    mockChannel.on.mock.calls[0][1]({ data: "fake", event: "event" })
    expect(called).toEqual("ab")

    unsubB()

    mockChannel.on.mock.calls[0][1]({ data: "fake", event: "event" })
    expect(called).toEqual("aba")

    unsubA()

    mockChannel.on.mock.calls[0][1]({ data: "fake", event: "event" })
    expect(called).toEqual("aba")
  })
})

describe("unbind", () => {
  it("works with no existing bindings", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    expect(subscription.bindings["test"]).toEqual(undefined)
    subscription.unbind("test")
    expect(subscription.bindings["test"]).toEqual([])
  })

  it("removes all current bindings for this event only", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    expect(subscription.bindings["test"]).toEqual(undefined)
    const unsubA = subscription.bind("test", () => (called += "a"))
    const unsubB = subscription.bind("test", () => (called += "b"))
    const other = subscription.bind("other", () => (called += "c"))
    expect(subscription.bindings["test"].length).toEqual(2)
    subscription.unbind("test")
    expect(subscription.bindings["test"]).toEqual([])
    expect(subscription.bindings["other"].length).toEqual(1)
  })
})

describe("unbindAll", () => {
  it("works with no existing bindings", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    expect(subscription.bindings).toEqual({})
    subscription.unbindAll()
    expect(subscription.bindings).toEqual({})
  })

  it("removes all current bindings", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    expect(subscription.bindings["test"]).toEqual(undefined)
    const unsubA = subscription.bind("test", () => (called += "a"))
    const unsubB = subscription.bind("other", () => (called += "b"))
    subscription.unbindAll()
    expect(subscription.bindings).toEqual({})
  })
})

describe("close", () => {
  function generateReceiveMock(eventToCallback) {
    const receiveMock = {
      receive: (event, fn) => {
        if (event === eventToCallback) {
          fn()
        }

        return receiveMock
      },
    }
    return receiveMock
  }

  it("resolves once closed", (done) => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const receiveMock = generateReceiveMock("ok")
    mockChannel.leave.mockReturnValue(receiveMock)

    subscription.close().then(done)
  })

  it("rejects if error", (done) => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const receiveMock = generateReceiveMock("error")
    mockChannel.leave.mockReturnValue(receiveMock)

    subscription.close().catch(done)
  })

  it("rejects if timeout", (done) => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const receiveMock = generateReceiveMock("timeout")
    mockChannel.leave.mockReturnValue(receiveMock)

    subscription.close().catch(done)
  })

  it("resolves without a channel", (done) => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.close().then(done)
  })
})

describe("hasBindings", () => {
  it("is false by default", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    expect(subscription.hasBindings()).toEqual(false)
  })

  it("is true with a binding", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const unsub = subscription.bind("other", () => (called += "b"))

    expect(subscription.hasBindings()).toEqual(true)
  })

  it("is false after unbinding", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const unsub = subscription.bind("other", () => (called += "b"))

    expect(subscription.hasBindings()).toEqual(true)

    unsub()

    expect(subscription.hasBindings()).toEqual(false)
  })

  it("is false after unbindAll", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const unsub = subscription.bind("other", () => (called += "b"))

    expect(subscription.hasBindings()).toEqual(true)

    subscription.unbindAll()

    expect(subscription.hasBindings()).toEqual(false)
  })

  it("is false after unbind/1", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const unsub = subscription.bind("other", () => (called += "b"))

    expect(subscription.hasBindings()).toEqual(true)

    subscription.unbind("other")

    expect(subscription.hasBindings()).toEqual(false)
  })

  it("is true after unbinding a single function but there's more", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const unsub = subscription.bind("other", () => (called += "b"))
    const unsubB = subscription.bind("other", () => (called += "b"))

    expect(subscription.hasBindings()).toEqual(true)

    unsub()

    expect(subscription.hasBindings()).toEqual(true)
  })

  it("is true after unbinding a single function but there's more in another event", () => {
    const subscription = new Subscription("chName", { pushEx: mockPushEx })
    subscription.setup()

    const unsub = subscription.bind("a", () => (called += "b"))
    const unsubB = subscription.bind("other", () => (called += "b"))

    expect(subscription.hasBindings()).toEqual(true)

    unsub()

    expect(subscription.hasBindings()).toEqual(true)

    unsubB()

    expect(subscription.hasBindings()).toEqual(false)
  })
})


src/pushex.js
import { Socket } from "phoenix"
import { Subscription } from "./subscription"

const NO_OP = () => {}

const DEFAULT_SOCKET_RECONNECT_ALGORITHM = tries => {
  return [3000, 6000, 10000, 20000][tries - 1] || 30000
}

export class Pushex {
  constructor(url, { getParams, onConnect, onConnectionError, socketReconnectAlgorithm }) {
    if (!url) {
      throw new Error("URL is not valid")
    }

    this.reconnectAlgorithm = socketReconnectAlgorithm || DEFAULT_SOCKET_RECONNECT_ALGORITHM
    this.getParams = getParams || (() => Promise.resolve({}))
    this.onConnect = onConnect || NO_OP
    this.onConnectionError = onConnectionError || NO_OP
    this.subscriptions = {}
    this._setupSocket(url)
  }

  connect() {
    return this.resetParams().then(() => {
      this.socket.connect()
    })
  }

  disconnect() {
    return new Promise(resolve => {
      this.socket.disconnect(() => resolve())
    })
  }

  resetParams() {
    return this.getParams().then(params => {
      this.socket.params = () => params
      return this
    })
  }

  getExistingSubscription(channelName) {
    return this.subscriptions[channelName]
  }

  subscribe(channelName) {
    if (!this.getExistingSubscription(channelName)) {
      const subscription = new Subscription(channelName, { pushEx: this })
      this.subscriptions[channelName] = subscription
    }

    return this.getExistingSubscription(channelName).setup()
  }

  unsubscribe(channelName) {
    return new Promise((resolve, reject) => {
      const subscription = this.subscriptions[channelName]

      if (subscription) {
        return subscription
          .close()
          .then(() => {
            delete this.subscriptions[channelName]
            resolve()
          })
          .catch(reject)
      } else {
        return resolve()
      }
    })
  }

  getSocket() {
    return this.socket
  }

  // private

  _setupSocket(url) {
    this.socket = new Socket(url, {
      reconnectAfterMs: this.reconnectAlgorithm
    })

    this.socket.onOpen(() => {
      this.onConnect(this)
    })

    this.socket.onError(() => {
      this.onConnectionError(this)
    })
  }
}


src/subscription.js
export class Subscription {
  constructor(channelName, { pushEx }) {
    this.channelName = channelName
    this.pushEx = pushEx
    this.bindings = {}
  }

  hasBindings() {
    return (
      Object.keys(this.bindings).filter(key => {
        return this.bindings[key].length > 0
      }).length > 0
    )
  }

  setup() {
    if (!this.channel) {
      this.channel = this.pushEx.getSocket().channel(this.channelName)
      this.channel.join()
      this.channel.on("msg", this._handleMessage.bind(this))
    }

    return this
  }

  close() {
    return new Promise((resolve, reject) => {
      if (this.channel) {
        this.channel
          .leave()
          .receive("ok", () => {
            this.channel = null
            resolve()
          })
          .receive("error", reject)
          .receive("timeout", reject)
      } else {
        resolve()
      }
    })
  }

  bind(eventName, fn) {
    this.bindings[eventName] = this.bindings[eventName] || []
    this.bindings[eventName].push(fn)

    return () => {
      this.bindings[eventName] = this.bindings[eventName].filter(testFn => testFn !== fn)
    }
  }

  unbind(eventName) {
    this.bindings[eventName] = []
  }

  unbindAll() {
    this.bindings = {}
  }

  // private

  _handleMessage({ data, event }) {
    this._runBindings(event, [data, event])
    this._runBindings("*", [data, event])
  }

  _runBindings(eventName, args) {
    const bindings = this.bindings[eventName] || []
    bindings.forEach(bindingFn => {
      this._invokeFunctionWithAsyncErrorHandling(bindingFn, args)
    })
  }

  _invokeFunctionWithAsyncErrorHandling(fn, args) {
    try {
      fn.apply(fn, args)
    } catch (e) {
      setTimeout(() => {
        throw e
      }, 0)
    }
  }
}


src/subscription.ts
import { Channel } from "phoenix"
import { Pushex } from "./pushex"

export type SubscriptionFunction<T = any> = (data: T, event: string) => void

export class Subscription {
  private channelName: string
  private pushEx: Pushex
  private bindings: Record<string, SubscriptionFunction[]>
  private channel?: Channel

  constructor(channelName: string, { pushEx }: { pushEx: Pushex }) {
    this.channelName = channelName
    this.pushEx = pushEx
    this.bindings = {}
  }

  public hasBindings() {
    return (
      Object.keys(this.bindings).filter((key) => {
        return this.bindings[key].length > 0
      }).length > 0
    )
  }

  public bind(eventName: string, fn: SubscriptionFunction) {
    this.bindings[eventName] = this.bindings[eventName] || []
    this.bindings[eventName].push(fn)

    return () => {
      this.bindings[eventName] = this.bindings[eventName].filter((testFn) => testFn !== fn)
    }
  }

  public unbind(eventName: string) {
    this.bindings[eventName] = []
  }

  public unbindAll() {
    this.bindings = {}
  }

  // Public, but for internal use
  public setup() {
    if (!this.channel) {
      this.channel = this.pushEx.getSocket().channel(this.channelName)
      this.channel.join()
      this.channel.on("msg", this.handleMessage.bind(this))
    }

    return this
  }

  // Public, but for internal use
  public close() {
    return new Promise((resolve, reject) => {
      if (this.channel) {
        this.channel
          .leave()
          .receive("ok", () => {
            this.channel = undefined
            resolve()
          })
          .receive("error", reject)
          .receive("timeout", reject)
      } else {
        resolve()
      }
    })
  }

  // private

  private handleMessage({ data, event }: { data: any; event: string }) {
    this.runBindings(event, [data, event])
    this.runBindings("*", [data, event])
  }

  private runBindings(eventName: string, args: [any, string]) {
    const bindings = this.bindings[eventName] || []
    bindings.forEach((bindingFn) => {
      this.invokeFunctionWithAsyncErrorHandling(bindingFn, args)
    })
  }

  private invokeFunctionWithAsyncErrorHandling(fn: SubscriptionFunction, args: [any, string]) {
    try {
      fn.apply(fn, args)
    } catch (e) {
      setTimeout(() => {
        throw e
      }, 0)
    }
  }
}


src/pushex.ts
import { Socket } from "phoenix"
import { Subscription } from "./subscription"

const NO_OP = () => {}

const DEFAULT_SOCKET_RECONNECT_ALGORITHM = (tries: number) => {
  return [3000, 6000, 10000, 20000][tries - 1] || 30000
}

type PushexConfig = {
  getParams: () => Promise<Record<string, any>>
  onConnect: (instance: Pushex) => void
  onConnectionError: (instance: Pushex) => void
  socketReconnectAlgorithm: typeof DEFAULT_SOCKET_RECONNECT_ALGORITHM
}

export class Pushex {
  private subscriptions: Record<string, Subscription>
  private socket!: Socket
  private config: PushexConfig

  constructor(url: string, config: PushexConfig) {
    if (!url) {
      throw new Error("URL is not valid")
    }

    this.config = {
      getParams: config.getParams || (() => Promise.resolve({})),
      onConnect: config.onConnect || NO_OP,
      onConnectionError: config.onConnectionError || NO_OP,
      socketReconnectAlgorithm: config.socketReconnectAlgorithm || DEFAULT_SOCKET_RECONNECT_ALGORITHM,
    }

    this.subscriptions = {}
    this.setupSocket(url)
  }

  public connect() {
    return this.resetParams().then(() => {
      this.socket.connect()
    })
  }

  public disconnect() {
    return new Promise((resolve) => {
      this.socket.disconnect(() => resolve())
    })
  }

  /**
   * This function exists in order to provide a promise interface to params. Phoenix's params closure
   * is less friendly when the parameters are provided asynchronously.
   */
  public resetParams() {
    return this.config.getParams().then((params) => {
      // @ts-ignore Acknowledged interaction with a private attribute, unlikely to change upstream
      this.socket.params = () => params
      return this
    })
  }

  public getExistingSubscription(channelName: string) {
    return this.subscriptions[channelName]
  }

  public subscribe(channelName: string) {
    if (!this.getExistingSubscription(channelName)) {
      const subscription = new Subscription(channelName, {
        pushEx: this,
      })

      this.subscriptions[channelName] = subscription
    }

    return this.getExistingSubscription(channelName).setup()
  }

  public unsubscribe(channelName: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const subscription = this.subscriptions[channelName]

      if (subscription) {
        return subscription
          .close()
          .then(() => {
            delete this.subscriptions[channelName]
            resolve()
          })
          .catch(reject)
      } else {
        return resolve()
      }
    })
  }

  public getSocket() {
    return this.socket
  }

  // private

  private setupSocket(url: string) {
    this.socket = new Socket(url, {
      reconnectAfterMs: this.config.socketReconnectAlgorithm,
    })

    this.socket.onOpen(() => {
      this.config.onConnect(this)
    })

    this.socket.onError(() => {
      this.config.onConnectionError(this)
    })
  }
}
